lib/spack/spack/solver/concretize.lp

```diff
@@ -1004,21 +958,15 @@ equality_attr("external_spec_selected").
 %   2. a `#minimize{ 0@2 : #true }.` statement that ensures the criterion
 %      is displayed (clingo doesn't display sums over empty sets by default)
 
-% Try to minimize the number of nodes of a given package
-% asp.py will generate statements of the following form for each literal "package"
-% indexed in an arbitrary but deterministic order.
-opt_criterion(300, "number of non-identical package nodes").
-#minimize { 0@300: #true }.
-#minimize { 1@300,PSID1,PSID2,Package : package_not_equal(PSID1, PSID2, Package), representative(PSID1, Package), representative(PSID2, Package) }.
-% asp.py will generate statements of a similar form for each literal "virtual"
-opt_criterion(301, "number of virtual providers").
-#minimize { 0@301: #true }.
-#minimize { 1@301,PSID1,PSID2,Virtual : virtual_not_equal(PSID1, PSID2, Virtual), representative_virtual(PSID1, Virtual), representative_virtual(PSID2, Virtual) }.
+% Minimize the number of PSID in the solution
+opt_criterion(300, "number of psids").
+#minimize { 0@300 : #true }.
+#minimize { 1@300,PSID : literal_process_space(_, PSID) }.
 
 % Try hard to reuse installed packages (i.e., minimize the number built)
 opt_criterion(100, "number of packages to build (vs. reuse)").
 #minimize { 0@100: #true }.
-#minimize { 1@100,Package,PSID : build_representative(PSID, Package), optimize_for_reuse() }.
+#minimize { 1@100,Package : build(PSID, Package), optimize_for_reuse() }.
```

Comment:
No, not really sure. I think you're right it should be `build_representative` because at best here I am double counting the same spec twice if it comes from different PSIDs. This means though that we can't get rid of the rules that are growing quadratically with PSIDs in grounding.

Reply:
Also, leaving `PSID` out in the lhs is a typo :slightly_frowning_face:

Reply:
@becker33 @tgamblin Another data point: concretizing [the pipelines](https://gitlab.spack.io/spack/spack/-/pipelines/98600) took an insane amount of time (and e4s didn't concretize in the 6hrs limit) I think the main issue might be we are introducing a lot of symmetry with PSIDs. Think of this simplified case of allowing clingo to place specs in `M` different PSIDs. Assume `N` of the specs can be moved around from a PSID to the other because they don't interfere with each other. That gives `M^N` possible optimal combinations.

Of course this is just rough intuition, since the reality of a solve is much more dynamic, but it would be a nice thing to try to introduce some artificial symmetry breaking rule in the PSID space and see if we benefit from it.